Feature/#5 naver id login

bsgreentea opened the pull request:
충돌 수정하고 다시 올리겠습니다

bsgreentea committed:
네아로 환경설정 #5

bsgreentea committed:
Style #5 버튼 생성
- 네아로 버튼 생성
- 네아로 설정값 추가

bsgreentea committed:
Fix #5 중복 세팅 수정
- 이미 네아로 세팅이 되어있었기 때문에 중복되는 부분 삭제
- 다시는 이런 뻘짓을 하지 않겠습니다

bsgreentea committed:
Style #5 액티비티 추가
- 로그인 액티비티(네아로 버튼 추가)
- 사용자 정보 액티비티(로그아웃과 탈퇴 버튼 추가)

bsgreentea committed:
Style #5 로그인, 사용자 정보 xml 수정

bsgreentea committed:
Style #5 스트링 추가 및 변경

bsgreentea committed:
Implement #5 네아로 매니저 구현
- 네이버 아이디로 로그인 매니저 구현
- 핸들러와 로그아웃 함수 구현

bsgreentea committed:
Implement #5 로그인, 로그아웃 구성
- 탈퇴 기능은 코루틴으로 바꿔보기
- 테스트 버튼은 추후에 삭제할 예정

ChoMk commented on app/src/main/AndroidManifest.xml:
```diff
@@ -1,23 +1,31 @@
+        android:usesCleartextTraffic="true">
         <activity android:name=".MainActivity"></activity>
-        <activity android:name=".ui.sample.SampleActivity">
+        <activity android:name=".LoginActivity">
```
ui 패키지 안에다가 넣으면 어떨지?

ChoMk commented on app/src/main/java/com/example/messenger/LoginActivity.kt:
```diff
@@ -0,0 +1,42 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_login)
+
+        init()
```
코틀린에 init이라는 함수가 있어서 함수명으로는 별로인거 같아

ChoMk replied:
생성자 만들 때 init사용

ChoMk commented on app/src/main/java/com/example/messenger/LoginActivity.kt:
```diff
@@ -0,0 +1,42 @@
+        init()
+        testButton = findViewById(R.id.test_button)
+        testButton.setOnClickListener {
+            startActivityForResult(Intent(this, UserInfoActivity::class.java), 1)
```
test이기는 하지만... start 로직을 아래와 같이 만들면 좋을 듯 나중에...
class LoginActionHelper(){
     val loginActivity
     fun startUserInfoActivity(){
           loginActivity.startactivity()
     }
}

저 위에 클레스를 현 클레스에서 초기화하고 이를 데이터바인딩을 통해서 testButton을 현 액티비티에서 초기화하지 않고 람다식으로 클릭하는 경우에 바인딩된 helper를 통하여 엑티비티 이동하도록하면 좋을 듯
activity안에서 저런식으로 하는게 잘못된건 아니지만... activity를 다른 모듈들과 연결하는 고리로 생각하면 어떨까하는 생각임...
그리고 이러한 start로직을 viewmodel로 분리하지 않은 이유는 context에 영향을 받지 않게하려고...(안드로이드 프레임워크에 영향을 받지 않는 로직으로 최대한 만들기 위하여?...)

ChoMk commented on app/src/main/java/com/example/messenger/NaverLoginManager.kt:
```diff
@@ -0,0 +1,53 @@
+
+        fun loginInstance(): OAuthLogin? {
+            if (loginInstance == null) {
+                loginInstance = OAuthLogin.getInstance()
```
loginInstance?.let{
}
이런식으로 사용하면 null 체크를 진행하고 다음으로 진행가능

ChoMk commented on app/src/main/java/com/example/messenger/NaverLoginManager.kt:
```diff
@@ -0,0 +1,53 @@
+        fun loginInstance(): OAuthLogin? {
+            if (loginInstance == null) {
+                loginInstance = OAuthLogin.getInstance()
+                loginInstance!!.init(
```
!!는 거의 쓰지 않는다고 함.. 쓰지 말라고 하고..

ChoMk commented on app/src/main/java/com/example/messenger/NaverLoginManager.kt:
```diff
@@ -0,0 +1,53 @@
+        }
+
+        fun naverLogout() {
+            loginInstance()!!.logout(context)
```
loginInstance()?.logout(context)이런식으로 진행하면 인스턴스가 있으면 다음을 진행함... 근데 만약에 어디서 인스턴스를 초기화하지 않은상태에서 저걸 실행하면 뻑남..

bsgreentea committed:
Merge branch 'develop' into feature/#5_naver_id_login

# Conflicts:
#	app/src/main/AndroidManifest.xml

ChoMk commented on app/src/main/java/com/example/messenger/NaverLoginManager.kt:
```diff
@@ -0,0 +1,53 @@
+            }
+        }
+
+    }
```
/*
 * TestNaverLoginManager.kt 2020. 5. 2
 *
 * Copyright 2020 BasicBug. All rights Reserved.
 *
 */

package com.example.messenger

import android.widget.Toast
import com.nhn.android.naverlogin.OAuthLogin
import com.nhn.android.naverlogin.OAuthLoginHandler

/**
 * @author MyeongKi
 */
object TestNaverLoginManager : OAuthLoginHandler() {

    private const val OAUTH_CLIENT_ID = "nJg0Pj0f2rEBCBQSZe6s"
    private const val OAUTH_CLIENT_SECRET = "2Ca8FHXsj4"
    private val context = MessengerApp.applicationContext()
    private var loginInstance = OAuthLogin.getInstance().apply {
        init(
            context,
            OAUTH_CLIENT_ID,
            OAUTH_CLIENT_SECRET,
            R.string.app_name.toString()
        )
    }


    fun naverLogout() {
        loginInstance?.logout(context)
    }

    override fun run(success: Boolean) {
        if (success) {
            val accessToken = loginInstance?.getAccessToken(context).toString()
            val refreshToken = loginInstance?.getRefreshToken(context).toString()
            val loginState = loginInstance?.getState(context).toString()
            Toast.makeText(context, "로그인되었습니다.", Toast.LENGTH_SHORT).show()
        } else {
            val errorCode: String = loginInstance?.getLastErrorCode(context)?.code ?: ""
            val descCode: String = loginInstance?.getLastErrorDesc(context) ?: ""
            Toast.makeText(context, "로그인에 실패했습니다.", Toast.LENGTH_SHORT).show()
        }
    }
}
-> 한번 수정해봄 메니저니깐 어차피 싱글톤으로 유지해도 될거 같아서 object로 진행했어

ChoMk commented on app/src/main/java/com/example/messenger/UserInfoActivity.kt:
```diff
@@ -0,0 +1,65 @@
+/**
+ * @author bsgreentea
+ */
+class UserInfoActivity : AppCompatActivity(), View.OnClickListener {
```
위에 언급했지만 리스너를 받고 xml을 정의해서 사용하는 것 보다는 엑션 처리를 도와주는 helper를 이용하거나 혹은 로직을 바로 갖고 갈 수 있는 viewmodel을 통하여 람다식으로 바인딩해주는게 좋을 듯

ChoMk commented on app/src/main/java/com/example/messenger/UserInfoActivity.kt:
```diff
@@ -0,0 +1,65 @@
+        init()
+    }
+
+    fun init() {
```
언급한 init 중복

ChoMk replied:
init{
}
이렇게 하면은 만들어지는 동시에 자바의 생성자와 같으 효과를 줄 수 있어

ChoMk commented on app/src/main/java/com/example/messenger/UserInfoActivity.kt:
```diff
@@ -0,0 +1,65 @@
+
+    class DeleteTokenTask : AsyncTask<Void, Void, Void>() {
+        override fun doInBackground(vararg p0: Void?): Void? {
+            loginInstance()!!.logoutAndDeleteToken(MessengerApp.applicationContext())
```
이부분은 rx로 못 만드나?

bsgreentea committed:
Implement #5 상수 추가

bsgreentea committed:
Implement #5 로그인 매니저 수정

bsgreentea commented on app/src/main/java/com/example/messenger/NaverLoginManager.kt:
```diff
@@ -1,53 +1,91 @@
+        Schedulers.newThread().createWorker().schedule {
+            loginInstance.logoutAndDeleteToken(context)
         }
+    }
```
주석 쳐놓은 Observable 부분을 사용하면 계속 탈퇴 실패가 되는데, 에러 코드를 확인해봐도 등록되지 않은 에러 코드가 나타납니다ㅜㅜ
탈퇴를 시도하면 로컬 토큰만 삭제가 되어서 로그아웃만 되는데, 사용법이 잘못된건가요?? 리액티브 잘 모르겠네요.. @ChoMk

bsgreentea committed:
Implement #5 데이터 바인딩 적용
- UserInfoActivity, LoginActivity 데이터바인딩 적용
- UserInfoActivity, LoginActivity 디렉토리 조정

bsgreentea committed:
Style #5 데이터바인딩 적용
- activity_login과 activity_user_info에 데이터바인딩 적용

bsgreentea committed:
Implement #5 로그인 헬퍼 구성
- 로그인, 탈퇴, 테스트 버튼 클릭에 대한 함수 구성

bsgreentea commented:
로그아웃, 탈퇴, 테스트(UserInfoActivity로 가는) 버튼에 데이터바인딩 적용해봤는데 확인 부탁드립니다. 작동은 제대로 하는데 맞게 한건진 잘 모르겠네요ㅎㅎ;

뷰모델이랑 같이 써보려고 했는데 이건 잘 모르겠어서 일단 이대로 진행했습니다.
뷰모델 활용이랑 disposable 사용 등 전체적인 흐름이 어떻게 되는건지 참교육이 필요합니다... 그냥 간단하게 쓰던 뷰모델이랑은 차원이 다르네요ㅜㅜ

탈퇴 기능은 아직 옵저버에 대해서 조금 더 공부해보고 다시 만들어보겠습니다.

ChoMk commented:
@bsgreentea 일단 잘 진행했으 ㅋㅋㅋㅋ helper랑 전반적인 구조는 따로 지금 만들고 있는 #10 이슈로 보여드림... 이 이슈가 생각보다 커져서...(전반적인 구조를 추가로 잡고 있어서...) 머지할겡
